File: tests/r_parity/80_contdid.R
# StatsPAI continuous-treatment DiD parity (R side) -- Module 80.
#
# Runs contdid::cont_did (Callaway, Goodman-Bacon & Sant'Anna's own package)
# on the panel dumped by 80_contdid.py, across three spline specifications.
#
# contdid is not on CRAN:
#   remotes::install_github("bcallaway11/contdid")
#
# Standard errors are not emitted: contdid routes them through the pte
# package's aggregation layer, which StatsPAI does not replicate.
#
# Tolerance: rel < 1e-6.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages(library(contdid))

MODULE <- "80_contdid"
GRID_POINTS <- c(1, 31, 61, 90)  # 1-based; the Python side uses 0, 30, 60, 89

df <- as.data.frame(read_csv_strict(MODULE))

rows <- list()
for (spec in list(c(1, 0), c(3, 0), c(3, 2))) {
  degree <- spec[1]
  num_knots <- spec[2]
  tag <- sprintf("d%dk%d", degree, num_knots)
  res <- suppressWarnings(cont_did(
    yname = "Y", tname = "time_period", idname = "id", dname = "D",
    data = df, gname = "G", target_parameter = "slope",
    aggregation = "dose", treatment_type = "continuous",
    dose_est_method = "parametric", control_group = "nevertreated",
    degree = degree, num_knots = num_knots, bstrap = FALSE, cband = FALSE
  ))
  for (k in seq_along(GRID_POINTS)) {
    j <- GRID_POINTS[k]
    rows[[length(rows) + 1]] <- parity_row(
      module = MODULE, statistic = sprintf("%s_att_d_%d", tag, j - 1),
      estimate = as.numeric(res$att.d[j]), se = NA_real_, n = nrow(df) / 2
    )
    rows[[length(rows) + 1]] <- parity_row(
      module = MODULE, statistic = sprintf("%s_acrt_d_%d", tag, j - 1),
      estimate = as.numeric(res$acrt.d[j]), se = NA_real_, n = nrow(df) / 2
    )
  }
  rows[[length(rows) + 1]] <- parity_row(
    module = MODULE, statistic = sprintf("%s_overall_att", tag),
    estimate = as.numeric(res$overall_att), se = NA_real_, n = nrow(df) / 2
  )
  rows[[length(rows) + 1]] <- parity_row(
    module = MODULE, statistic = sprintf("%s_overall_acrt", tag),
    estimate = as.numeric(res$overall_acrt), se = NA_real_, n = nrow(df) / 2
  )
}

write_results(MODULE, rows, extra = list(
  reference = "contdid::cont_did",
  contdid = as.character(utils::packageVersion("contdid"))
))

File: tests/r_parity/80_contdid.py
"""StatsPAI continuous-treatment DiD parity (Python side) -- Module 80.

Runs ``sp.cgs_continuous_did`` on a two-period panel with a continuous dose;
the companion R script runs ``contdid::cont_did`` (Callaway, Goodman-Bacon &
Sant'Anna's own package) on the same dumped CSV.

Three spline specifications are emitted -- linear with no interior knots,
cubic with none, cubic with two -- because the whole estimator is a spline
regression in the dose and a single degree would leave the basis mostly
unexercised.

``curve_basis="reference"`` is used for the dose curves. ``contdid`` fits the
spline on the range of the observed treated doses but evaluates the reported
curves on a basis re-anchored to the ends of the dose GRID, so its reported
curves are a rescaled version of the fitted dose response and do not line up
with the overall ACRT it returns alongside them. StatsPAI defaults to one
consistent basis and keeps the reference's behaviour behind that flag. The overall ATT and ACRT
are computed on the fitted basis in both packages and need no flag.

Standard errors are not compared: ``contdid`` routes them through the ``pte``
package's aggregation layer, which StatsPAI does not replicate.

Tolerance: rel < 1e-6 on the curves and both overall quantities.
"""

from __future__ import annotations

import warnings

import numpy as np
import pandas as pd

from _common import ParityRecord, dump_csv, write_results

MODULE = "80_contdid"
SPECS = ((1, 0), (3, 0), (3, 2))
GRID_POINTS = (0, 30, 60, 89)  # where on the 90-point grid to compare


def _panel() -> pd.DataFrame:
    """Two periods, half the units dosed continuously, a quadratic response.

    The quadratic term is what makes ACRT vary with the dose; a purely
    linear design would let a wrong basis pass unnoticed.
    """
    rng = np.random.default_rng(19)
    rows = []
    for uid in range(1, 2001):
        treated = uid <= 1000
        d = float(rng.uniform(0.02, 1.0)) if treated else 0.0
        fe = rng.normal(0.0, 1.0)
        for t in (1, 2):
            eff = (1.6 * d + 0.9 * d**2) if (treated and t == 2) else 0.0
            rows.append(
                {
                    "id": uid,
                    "time_period": t,
                    "G": 2 if treated else 0,
                    "D": d,
                    "Y": fe + 0.5 * t + eff + rng.normal(0, 0.5),
                }
            )
    return pd.DataFrame(rows)


def main() -> None:
    import statspai as sp

    df = _panel()
    dump_csv(df, MODULE)

    rows: list[ParityRecord] = []
    for degree, num_knots in SPECS:
        tag = f"d{degree}k{num_knots}"
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            res = sp.cgs_continuous_did(
                df,
                y="Y",
                dose="D",
                time="time_period",
                unit="id",
                cohort="G",
                degree=degree,
                num_knots=num_knots,
                curve_basis="reference",
            )
        for j in GRID_POINTS:
            rows.append(
                ParityRecord(
                    module=MODULE,
                    side="py",
                    statistic=f"{tag}_att_d_{j}",
                    estimate=float(res.att_d[j]),
                    n=int(res.n_units),
                )
            )
            rows.append(
                ParityRecord(
                    module=MODULE,
                    side="py",
                    statistic=f"{tag}_acrt_d_{j}",
                    estimate=float(res.acrt_d[j]),
                    n=int(res.n_units),
                )
            )
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"{tag}_overall_att",
                estimate=float(res.overall_att),
                n=int(res.n_units),
            )
        )
        rows.append(
            ParityRecord(
                module=MODULE,
                side="py",
                statistic=f"{tag}_overall_acrt",
                estimate=float(res.overall_acrt),
                n=int(res.n_units),
            )
        )

    write_results(MODULE, "py", rows, extra={"estimator": "cgs_continuous_did"})


if __name__ == "__main__":
    main()
